### ISSUE TITLE
fp16 causes loss to be nan?

### ISSUE THREAD

**Issue by vedantroy** [lucidrains/imagen-pytorch#129]
Hi! When using the fp16 option, my loss becomes nan.
I'm using a V100.
Is there any other option I need to configure besides `fp16=True` in the trainer?

**Comment by jacobwjs** [comment]
I'm training 100's of thousands of steps using fp16. No nan's to speak of. Perhaps look at gradients, and if there's an issue clip https://github.com/lucidrains/imagen-pytorch/blob/e1a8dc258246bae465d489754afacea8dd9bbf43/imagen_pytorch/trainer.py#L198

**Comment by lucidrains** [comment]
@vedantroy Hi Vedant again :wave: 

Could you keep your current settings, and try turning https://github.com/lucidrains/imagen-pytorch/commit/bdb6ee0fc06404fb82d97b9590c2830fdce5f27b#diff-edef3c5fe92797a22c0b8fc6cca1d57b4b84ef03dfdfe802ed9147e21fc88109R1036 on?

**Comment by lucidrains** [comment]
in other words, you are my guinea pigs haha

**Comment by lucidrains** [comment]
basically if we prove out this approach, no more attention overflow problems with trivial modifications

**Comment by jacobwjs** [comment]
very cool! can you think of a metric that with and without it would give you a good idea if it's beneficial.  basically how do I test this and know if it has a positive impact?

**Comment by lucidrains** [comment]
@jacobwjs we just need comparison runs with that flag turned on vs off!

**Comment by lucidrains** [comment]
@jacobwjs my thoughts are, even if cosine sim attention lags a bit behind performance wise, as long as we quantify that, it may be a good tradeoff (as long as we get testimony that it alleviates or greatly reduces overflows)

for example, the transformer field have largely moved onto pre-layernorm for stability reasons, even though it is a bit behind the traditional post-layernorm

**Comment by lucidrains** [comment]
@jacobwjs thank you Jacob! :pray:

**Comment by vedantroy** [comment]
@lucidrains running on version 1.1.2 of imagen_pytorch with the following configuration:

![image](https://user-images.githubusercontent.com/8407819/179616662-528db13e-0b52-407b-a686-a1f18351933e.png)

where
- params=256
- fp16=True

The loss is still nan. I'll try gradient clipping next.

**Comment by vedantroy** [comment]
@jacobwjs Can you post a pic of your config? Maybe I'm doing some super wrong.

**Comment by lucidrains** [comment]
i'm not seeing any issues, at least with unconditional training with batch size 32 and learning rate `1e-4`

**Comment by ryakr** [comment]
I have been having this same issue starting all the way back when it was first added during the accelerate addition, I thought it was just me! Running on a 3090 using text and images with any batch size, normal training is fine but if you use FP16 the training will not progress (images stay noisy and at whatever learning level they were) and loss will be nan. I even had my models fail to save silently, but have not tested that aspect on the newer versions.

**Comment by lucidrains** [comment]
try increasing the batch size to 32, then use `max_batch_size` of 8 or 4 to keep it from OOM

**Comment by vedantroy** [comment]
> try increasing the batch size to 32, then use `max_batch_size` of 8 or 4 to keep it from OOM

That worked! Would love if you gave an explanation of what is going on here. I assumed `max_batch_size` was just splitting the input into smaller batches underneath the hood, so the end result would be the same?

**Comment by lucidrains** [comment]
> > try increasing the batch size to 32, then use `max_batch_size` of 8 or 4 to keep it from OOM
> 
> That worked! Would love if you gave an explanation of what is going on here. I assumed `max_batch_size` was just splitting the input into smaller batches underneath the hood, so the end result would be the same?

yup, max batch size just makes sure it never processes more than that size at any time, preventing the memory from spilling over the limit!

**Comment by lucidrains** [comment]
@vedantroy just so i calibrate things in my head, is your current run using `cosine_sim_attn`?

**Comment by vedantroy** [comment]
> @vedantroy at minimum it should be 16 or 32

Got it, so what is going on here is that I'm updating the gradients every 32 samples, but processing data in batches of 8.

And no, I'm not using `cosine_sim_attn`. I'll restart the run now with that enabled.

**Comment by lucidrains** [comment]
@vedantroy ok! if you run into NaNs later on, try turning that on

if it works even with it off, then it is fine, you can just keep it that way

**Comment by vedantroy** [comment]
@lucidrains 

> @vedantroy ok! if you run into NaNs later on, try turning that on
> 
> if it works even with it off, then it is fine, you can just keep it that way

I'm happy to turn `cosine_sim_attn=True` if there's some experiment you want to run. But I'm guessing the experiment requires me hitting NaNs first?

**Comment by lucidrains** [comment]
yup, only required if you hit further instability!

**Comment by vedantroy** [comment]
> yup, only required if you hit further instability!

Nevermind, I'm running into the issue. I think I thought I wasn't running into it?

My config is:

```
{
    "unets": [
        {
            "type": "BaseUnet64",
            "image_size": 64,
            "params": {
                "dim": 256,
                "cosine_sim_attn": true
            }
        }
    ],
    "imagen": {
        "auto_normalize_img": true
    },
    "trainer": {
        "fp16": true
    },
    "imagen_pytorch_version": "1.1.2"
}
```

with a `max_batch_size=12` and a gradient update size (not sure what the really term is) of 48. Using a smaller `max_batch_size=2` does not fix the problem either. Turning `cosine_sim_attn` on/off doesn't change the issue either.

I'll try changing the gradient update size to 256 and see if that fixes the issue.

**Comment by lucidrains** [comment]
@vedantroy doh, yea, i think your code looks ok

do you want to try v1.1.5? https://github.com/lucidrains/imagen-pytorch/commit/5fca6872dad964bc20f6213aec267487cafc1a77

**Comment by vedantroy** [comment]
> @vedantroy doh, yea, i think your code looks ok
> 
> do you want to try v1.1.5? [5fca687](https://github.com/lucidrains/imagen-pytorch/commit/5fca6872dad964bc20f6213aec267487cafc1a77)

Hm, that didn't fix the issue. :/
Would love to assist with debugging but not sure how.
If you want, I can DM you ssh access to my machine?

**Comment by lucidrains** [comment]
@vedantroy sure, that'd be great!

**Comment by lucidrains** [comment]
@vedantroy one last thing to try would be to set a warmup for the learning rate https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py#L201

you can try 10000 steps for starters

**Comment by vedantroy** [comment]
> @vedantroy one last thing to try would be to set a warmup for the learning rate https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/trainer.py#L201
> 
> you can try 10000 steps for starters

Hm, just tried that didn't work. Sent a DM with my SSH creds.

**Comment by lucidrains** [comment]
@vedantroy thanks for pressing on this! i believe i have found the problem and the fix is in 1.1.6

**Comment by vedantroy** [comment]
> @vedantroy thanks for pressing on this! i believe i have found the problem and the fix is in 1.1.6

Thanks! This fixed the issue. Looking at the diff, it seems like you turn off fp16 when running the t5 model. I'm curious about what the intuition is behind this fix? I.e., my assumption is that if you do this the embeddings/masks will both be float32. But, presumably the rest of the model will be fp16, which will cause some issues right?

**Comment by lucidrains** [comment]
@vedantroy yea, pytorch autocast will take care of converting the types between the boundaries of code

**Comment by lucidrains** [comment]
@vedantroy you should let gwern know that you are training on danbooru :laughing:
